usr/sbin/pkcstok_migrate/pkcstok_migrate.c

```diff
@@ -0,0 +1,2367 @@
+        printf("ok, let's quit.\n");
+        goto done;
+    }
+
+    /* Get the SO pin to authorize migration */
+    if (!sopin) {
+        printf("Enter the SO PIN: ");
+        fflush(stdout);
+        ret = get_pin(&sopin, &sopinlen);
+        if (ret != 0) {
+            printf("Could not get SO PIN.\n");
+            goto done;
+        }
+    }
+
+    /* Get the USER pin to authorize migration */
+    if (!userpin) {
+        printf("Enter the USER PIN: ");
+        fflush(stdout);
+        ret = get_pin(&userpin, &userpinlen);
+        if (ret != 0) {
+            printf("Could not get USER PIN.\n");
+            goto done;
+        }
+    }
+
+    /* Verify the SO and USER PINs entered against NVTOK.DAT. */
+    ret = verify_pins(data_store, sopin, sopinlen, userpin, userpinlen);
+    if (ret) {
+        printf("Could not verify pins.\n");
+        goto done;
+    }
+
+    /* Check if data store is already new */
+    ret = datastore_is_312(data_store, sopin, userpin, &new);
+    if (ret == 0 && new) {
+        printf("Data store %s is already in new format.\n", data_store);
+        goto done;
+    }
+
+    /* Backup repository if not already done */
+    ret = backup_repository(data_store);
+    if (ret != CKR_OK) {
+        printf("Failed to create backup.\n");
+        goto done;
+    }
+
+    /* Perform all actions on the backup */
+    char data_store_new[PATH_MAX];
+    data_store_old = data_store;
+    snprintf(data_store_new, PATH_MAX, "%s_PKCSTOK_MIGRATE_BACKUP", data_store_old);
+
+    /* Create new temp token keys, which exist in parallel to the old ones
+     * until the migration is fully completed. */
+    if (!datastore_312_exists(data_store_new)) {
+        ret = create_temp_token_keys_312(data_store_new, sopin, userpin);
+        if (ret != CKR_OK) {
+            printf("Failed to create new token keys.\n");
+            goto done;
+        }
+    }
+
+    /* Migrate repository */
+    ret = migrate_repository(data_store_new, sopin, userpin);
+    if (ret != CKR_OK) {
+        printf("Failed to migrate repository.\n");
+        goto done;
+    }
+
+    /* Switch to new repository */
+    ret = switch_to_new_repository(data_store_old, data_store_new);
+    if (ret != CKR_OK) {
+        printf("Switch to new repository failed.\n");
+        goto done;
+    }
+
+    /* Now insert new 'tokversion=3.12' parm in opencryptoki.conf */
+    ret = update_opencryptoki_conf(slot_id, conf_dir);
+    if (ret != CKR_OK) {
+        printf("Failed to update opencryptoki.conf, you must do this manually.\n");
+        goto done;
+    }
+
+    printf("Please manually remove backups of:\n");
+    printf("  old data store %s\n", data_store_old);
+    printf("  old opencryptoki.conf in %s\n",conf_dir);
+
+    ret = CKR_OK;
+
+done:
+
+    free(sopin);
+    free(userpin);
+    free(data_store);
+    free(conf_dir);
+    free(verbose);
+    free(dll_name);
+
+    printf("pkcstok_migrate finished.\n");
+
+    return ret;
```

Comment:
The return code of main should probably be EXIT_SUCCESS or EXIT_FAILURE, not the OCK return code.

Reply:
ok